core/drivers/clk/clk_dt.c

```diff
@@ -0,0 +1,364 @@
+
+		clock_cells = fdt_clock_cells(fdt, parent_node);
+		idx += sizeof(phandle) + clock_cells * sizeof(uint32_t);
+	}
+}
+
+static void clk_setup_compatible(const void *fdt, const char *compatible,
+				 const struct clk_driver *clk_drv)
+{
+	int status = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	int node = fdt_node_offset_by_compatible(fdt, -1, compatible);
+
+	if (node < 0 || clk_get_provider(node))
+		return;
+
+	while (node >= 0) {
+		status = _fdt_get_status(fdt, node);
+		if (!(status & DT_STATUS_OK_SEC))
+			continue;
+
+		parse_clock_property(fdt, node);
+
+		res = clk_drv->setup(fdt, node);
+		if (res != TEE_SUCCESS) {
+			EMSG("Failed to setup clock driver for compatible %s",
+			      compatible);
+			panic();
+		}
+
+		node = fdt_node_offset_by_compatible(fdt, node, compatible);
+	};
+}
+
+static bool clk_dt_set_parent(struct clk *clk, struct clk *parent)
+{
+	size_t i = 0;
+
+	for (i = 0; i < clk_get_num_parents(clk); i++) {
+		if (clk_get_parent_by_index(clk, i) == parent) {
+			if (clk_set_parent(clk, i) != TEE_SUCCESS)
+				panic();
+			return true;
+		}
+	}
+
+	EMSG("Clock %s is not a parent of clock %s", parent->name, clk->name);
+	return false;
+}
+
+static void parse_assigned_clock(const void *fdt, int nodeoffset)
+{
+	int rate_len = 0;
+	int clock_idx = 0;
+	struct clk *clk = NULL;
+	unsigned long rate = 0;
+	struct clk *parent = NULL;
+	const uint32_t *rate_prop = NULL;
+
+	rate_prop = fdt_getprop(fdt, nodeoffset, "assigned-clock-rates",
+				&rate_len);
+	rate_len /= sizeof(uint32_t);
+
+	while (1) {
+		clk = clk_dt_get_by_idx_prop("assigned-clocks", fdt, nodeoffset,
+					     clock_idx);
+		if (!clk)
+			return;
+
+		parent = clk_dt_get_by_idx_prop("assigned-clock-parents", fdt,
+						nodeoffset, clock_idx);
+		if (parent) {
+			if (!clk_dt_set_parent(clk, parent))
+				return;
+		}
+
+		if (rate_prop && clock_idx <= rate_len) {
+			rate = fdt32_to_cpu(rate_prop[clock_idx]);
+			if (rate) {
+				if (clk_set_rate(clk, rate) != TEE_SUCCESS)
+					panic();
+			}
+		}
+
+		clock_idx++;
+	}
+}
+
+static void clk_probe_assigned(const void *fdt, int parent_node)
+{
+	int len = 0;
+	int child = 0;
+	int status = 0;
+
+	fdt_for_each_subnode(child, fdt, parent_node) {
+		clk_probe_assigned(fdt, child);
+
+		status = _fdt_get_status(fdt, child);
+		if (!status)
+			continue;
+
+		if (fdt_getprop(fdt, child, "assigned-clocks", &len))
+			parse_assigned_clock(fdt, child);
+	}
+}
+
+static TEE_Result clk_dt_probe(void)
+{
+	const struct dt_driver *drv = NULL;
+	const void *fdt = get_embedded_dt();
+	const struct dt_device_match *dm = NULL;
+	const struct clk_driver *clk_drv = NULL;
+
+	DMSG("Probing clocks from devicetree");
+	if (!fdt)
+		panic();
+
+	for_each_dt_driver(drv) {
+		if (drv->type != DT_DRIVER_CLK)
+			continue;
+
+		clk_drv = (const struct clk_driver *)drv->driver;
+		for (dm = drv->match_table; dm && dm->compatible; dm++)
+			clk_setup_compatible(fdt, dm->compatible, clk_drv);
+	}
+
+	clk_probe_assigned(fdt, -1);
+
```

Comment:
once clock drivers are all probed, can't we free the memory allocate for clock provider registering, `struct clk_dt_provider`? I think there are not required anymore, are they?

Reply:
Actually, the provider are still used for `clk_dt_get_by_idx_prop` which is used by drivers to query clocks from device tree. However, providers could probably be freed at the end of OP-TEE boot with a `boot_final` initcall such as `release_external_dt`. I don't think we can expect drivers to query clocks after that.

Reply:
agree